infra/pySourceCredChart/js/lib/pySourceCredChart.js

```diff
@@ -0,0 +1,320 @@
+				.attr("refY", 0)
+				.attr("markerWidth", 5)
+				.attr("markerHeight", 5)
+				.attr("orient", "auto")
+				.append("svg:path")
+				.attr("d", "M 0,-5 L 10 ,0 L 0,5");   
+		
+		//chart elements    
+		that.edge = that.chart.append("g");
+		that.node = that.chart.append("g");
+		
+		that.tooltip = that.chartContainer.append("div")
+			.attr("class", "toolTip")
+			.style('display', 'none')
+			.style('position', 'absolute')
+			.style('min-width' , '50px')
+			.style('height', 'auto')
+			.style('background', 'none repeat scroll 0 0 #ffffff');
+	}
+	
+	function d3updateNodes(that, nodes){
+		// node data join
+		// TODO: add drag behavior
+		
+		var nodeColor = d3.scaleOrdinal(d3.schemeCategory10);
+		var node = that.node.selectAll(".node").data(nodes);
+
+		// node exit	
+		node.exit()
+			.transition()
+			.ease(d3.easeQuad)
+			.duration(1000)
+			.remove();
+			
+		//node enter
+		var newNode = node.enter()
+			.append('circle')
+			.attr('class', 'node')
+			.on('mouseover', mouseOver)
+			.on('mouseout', mouseOff)
+			.on('click', clickHalo);
+       
+		// node update	
+		node.merge(newNode)
+			.transition()
+			.ease(d3.easeQuad)
+			.duration(1000)
+			.attr('fill', function(d){
+				if(that.mapping.nodeGroup){
+					return nodeColor(d[that.mapping.nodeGroup]);
+				} else {
+					return 'steelblue';
+				}
+			})
+			.attr('r', function(d){
+				if(that.mapping.nodeSize){
+					return Math.min(75,Math.max(1, that.scores[d.index] * 1000));
+				} else {
+					return 5;
+				}
+			});
+			
+		function mouseOver(){
+			var data = d3.select(this).data()[0];
+			var textDisplay = data['index'] + '<br>' + data[that.mapping.nodeLabel] + '<br>' + data[3] + '<br>' + data[4] ;
+			var nodeColor = d3.select(this).style('fill');
+			console.log(data);
+			
+			that.tooltip
+			  .style("left", (d3.event.pageX - 250) + 'px')
+			  .style("top", 0 + 'px')
+			  .style("display", "inline")
+			  .style("background", nodeColor)
+			  .html(function(){
+				  return textDisplay;
+			  });
+		}
+		
+		function mouseOff(){
+			that.tooltip.style("display", "none");
+		}
+		
+		function clickHalo(){
+			//reset classes
+			that.svg.selectAll('circle').attr('class', 'node');
+			that.svg.selectAll('line').attr('class', 'edge');
+			
+			//get node index
+			var nodeIdx = d3.select(this).data()[0].index;
+			console.log(nodeIdx);
+			
+			//change node class to halo
+			d3.select(this).attr('class', 'halo');
+			
+			//change connected links to class halo
+			console.log(that.svg.selectAll(".edge").data()[0]);
+			var links = that.svg.selectAll(".edge").filter(function(d){
+				console.log(d);
+                return (d.dstIndex == nodeIdx | d.srcIndex == nodeIdx);
+              });
+			links.attr('class', 'halo');
+			console.log(links);
+		}
+	}
+	
+	function d3updateEdges(that, links){
+		
+		var edgeColor = d3.scaleOrdinal(d3.schemeCategory10);
+		// edge data join
+		var edge = that.edge.selectAll('.edge').data(links);
+		
+		// edge exit
```

Comment:
same suggestion as with the other comments that don't provide any context except for what's already in the code

Reply:
Oops, missed this one.

Reply:
Should be cleaned up now.